tls/s2n_tls13_secrets.c

```diff
@@ -168,6 +168,10 @@ static S2N_RESULT s2n_derive_secret_with_context(struct s2n_connection *conn,
         s2n_extract_secret_type_t input_secret_type, const struct s2n_blob *label, message_type_t transcript_end_msg,
         struct s2n_blob *output)
 {
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(label);
+    RESULT_ENSURE_REF(output);
+
     RESULT_ENSURE(CONN_SECRETS(conn).extract_secret_type == input_secret_type, S2N_ERR_SECRET_SCHEDULE_STATE);
     RESULT_ENSURE(s2n_conn_get_current_message_type(conn) == transcript_end_msg, S2N_ERR_SECRET_SCHEDULE_STATE);
     RESULT_GUARD(s2n_derive_secret(CONN_HMAC_ALG(conn), &CONN_SECRET(conn, extract_secret),
```

Comment:
Technically `CONN_HMAC_ALG` dereferences `conn->secure` and `conn->secure->cipher_suite`. But maybe this amount of checking is unnecessary?

Reply:
@lrstewart ?

Reply:
Maybe that check should be part of `CONN_HMAC_ALG`?

Perhaps we should investigate to what extent these checks get eliminated at compile time. -- Such an investigation might cause us to be more confident that we can always add more checks.

Reply:
https://github.com/harrisonkaiser/s2n-tls/pull/1

Reply:
I mean it's necessary for our aggressively redundant checks, but it's hard to argue any of this is _necessary_ necessary :) But regardless, I wouldn't split that change into a separate PR.

I'm not sure adding a new macro works any better than raw ENSURE_REF checks. Someone modifying the code or adding a new call to CONN_HMAC_ALG still might not make the connection that they need that macro, because CONN_HMAC_ALG isn't obviously dereferencing anything.

I think your options are:
1. No additional checks. This is probably fine-- we null-check `conn->secure` and `conn->secure->cipher_suite` all over the place, and in reality they should never be able to be NULL anyway.
2. Raw ensure_ref checks. Could be equivalent to 1 if future devs remove the checks.
3. Your new macro. Could be equivalent to 1 if future devs remove the checks, but that seems less likely to happen.
4. Add the raw checks / macro, but only to the methods accessible from outside this module: https://github.com/harrisonkaiser/s2n-tls/blob/9cb1a6558479ec83304b1b3e2805504cec835fa6/tls/s2n_tls13_secrets.h#L46-L54 So even if future devs would have forgotten the checks, something higher is still checking. This is the approach the extensions framework takes.

I would be okay with any of those options.

By the way-- we have a much worse and more pervasive version of this problem with s2n_stuffer_data_available(): https://github.com/aws/s2n-tls/blob/main/stuffer/s2n_stuffer.h#L59-L62 It's not even capitalized to indicate that it's a macro :(

Reply:
I'm happy with option number 4 for now.

But your pointing out that this is a more pervasive issue leads me to suggest a solution that would be outside the scope of this PR: statement expressions.

Currently our compiler infrastructure doesn't support them. But they would enable us to solve this issue. For example we could replace `CONN_HMAC_ALG` with the following definition:

```c
#define CONN_HMAC_ALG(conn) ({                                      \
    RESULT_ENSURE_REF(conn);                                             \
    RESULT_ENSURE_REF((conn)->secure);                           \
    RESULT_ENSURE_REF((conn)->secure->cipher_suite);   \
    (conn)->secure->cipher_suite->prf_alg;})
```

The result would be a macro that couldn't be miss used.

I don't know what the constraints are restricting us from using statement expressions (I believes they are a long standing and well supported feature -- although definitely non-standard.)

Reply:
We do that where possible, but it has limitations:
* If you do that, you can't use the macro as an argument to a function. That's the primary use of CONN_HMAC_ALG, and a common use case for s2n_stuffer_data_available.
* That macro you proposed would only work for arguments that return s2n_result. You would then also need to generate a version that worked for int and null returns.

Reply:
Actually I'm fairly certain that statement expressions are allowed to be function arguments. 

This will compile, for instance:
```c
#include <stdio.h>

#define FOO(v) ({ if (v == 'a') { return -1; }; v;})

int f(int g) {
	return g;
}

int main() {
	while (a != -1) {
		int a = f(FOO(getchar()));
		printf("Got %c\n", (char)a);
	}
        return 0;
}
```

On the other hand -- yeah it does specialize the macro to functions that return s2n_result. (But maybe that just gives us a reason to make more functions return s2n_result?)

Reply:
Under the current framework we get a warning that becomes an error. This is because `--pedantic` warns that statement expressions are not C99 standard:
```
file.c:11:14: warning: use of GNU statement expression extension [-Wgnu-statement-expression]
                char a = f(FOO(getchar()));
                           ^
file.c:3:17: note: expanded from macro 'FOO'
#define FOO(v) ({ if (v == 'a') { return -1; }; v;})
                ^
1 warning generated.
```

Reply:
We would need to add `-Wno-gnu-statement-expression` to our command line options in the build system.

Reply:
I'm unfamiliar with statement expressions-- I didn't realize they were allowed to call return! That seems like it could lead to some hard to follow logic and unexpected behavior.

But it's not in the standard, and I see no reason to deviate from the standard just for more null checks.

Reply:
Ok -- I'll just implement 4 as described above.